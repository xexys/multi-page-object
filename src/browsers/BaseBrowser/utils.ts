import {PageElement} from '../../pageElements';

/**
 * Рекурсивно в глубину обходит объекты и массивы, и маппит значения PageElement в PageElement['value'].
 */
export const prepareExecuteCallbackArgs = (value: any): any => {
    if (value instanceof PageElement) {
        return value.value;
    }

    if (Array.isArray(value)) {
        return value.map(prepareExecuteCallbackArgs);
    }

    if (typeof value === 'object' && value !== null) {
        const keys = Object.keys(value);
        const obj: Record<string, any> = {};

        for (const key of keys) {
            obj[key] = prepareExecuteCallbackArgs(value[key]);
        }

        return obj;
    }

    return value;
};
