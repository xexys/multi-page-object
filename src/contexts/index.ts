export * from './TestContext';
export * from './CatTestContext';
export * from './E2ETestContext';
export * from './HermioneTestContext';
export * from './TestamentTestContext';
export * from './UnitTestContext';
export * from './utils';
