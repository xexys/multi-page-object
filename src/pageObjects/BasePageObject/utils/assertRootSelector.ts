import assert from 'assert';

import type {BasePageObject} from '../index';

export const assertRootSelector = ({rootSelector, name}: typeof BasePageObject): void => {
    assert(rootSelector, `Value '${rootSelector}' of '${name}.rootSelector' must not be empty!`);
    assert(
        typeof rootSelector === 'string',
        `Value '${rootSelector}' of '${name}.rootSelector' must be valid CSS selector string!`,
    );
};
