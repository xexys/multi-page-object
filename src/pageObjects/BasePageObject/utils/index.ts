export * from './assertRootSelector';
export * from './isPageObjectRootElement';
export * from './processArgs';
export * from './searchRootElement';
