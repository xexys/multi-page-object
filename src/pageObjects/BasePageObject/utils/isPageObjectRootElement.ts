import type {PageElement} from '../../../pageElements';
import {isPageElementMatchWithSelector} from '../../../pageElements';
import type {BasePageObject, GetPageObjectPageElement, PageObjectCtor} from '../index';

/**
 * В TS нельзя писать асинхронные гарды, но есть весьма лаконичный воркэраунд для таких ситуаций.
 * @see https://github.com/microsoft/TypeScript/issues/37681
 */
export const isPageObjectRootElement = async <T extends BasePageObject, E extends GetPageObjectPageElement<T>>(
    elem: PageElement,
    Ctor: PageObjectCtor<T>,
): Promise<E | null> => {
    // console.debug(elem);
    // console.debug(Ctor.rootSelector);
    const isMatch = await isPageElementMatchWithSelector(elem, Ctor.rootSelector);

    return isMatch ? (elem as E) : null;
};
