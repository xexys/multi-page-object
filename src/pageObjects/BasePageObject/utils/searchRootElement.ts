import type {TestContext} from '../../../contexts';
import {
    RootElementNotFoundError,
    createGetElementByIndexLogMessage,
    createGetElementByMapperLogMessage,
    createGetElementByQuerySelectorLogMessage,
    createGetElementByQuerySelectorAllLogMessage,
} from '../../../errors';
import type {PageElement, SearchPageElement} from '../../../pageElements';
import type {Locator, RootElementIndex, RootElementSelector, RootElementMapper} from '../../../locators';

const createQuerySelector = (selectors: string[]): string => {
    return selectors.join(' ');
};

/**
 * Поиск root-элемента через переданный список локаторов.
 * Локаторы применяются последовательно к результату поиска предыдущего локатора.
 *
 * @see BasePageObject._getLocators
 */
export const searchRootElement = async (context: TestContext, locators: Locator[]): Promise<PageElement> => {
    const searchLog: string[] = [];

    const _searchRootElement = async (
        _locators: Array<RootElementIndex | RootElementSelector | RootElementMapper>,
        container: SearchPageElement,
    ): Promise<PageElement | null> => {
        const selectors = [];

        for (let i = 0; i < _locators.length; i++) {
            const locator = _locators[i];

            if (typeof locator === 'string') {
                selectors.push(locator.trim());
            } else {
                const query = createQuerySelector(selectors);
                const elems = await container.$$(query);

                searchLog.push(createGetElementByQuerySelectorAllLogMessage({container, query, result: elems}));

                let root: PageElement | null | undefined;

                if (typeof locator === 'number') {
                    root = elems[locator];
                    searchLog.push(createGetElementByIndexLogMessage({index: locator, result: root}));
                } else {
                    root = await locator(context, elems);
                    searchLog.push(createGetElementByMapperLogMessage({mapper: locator, result: root}));
                }

                const restLocators = _locators.slice(i + 1);

                if (root && restLocators.length) {
                    return _searchRootElement(restLocators, root);
                }

                return root ?? null;
            }
        }

        if (selectors.length) {
            const query = createQuerySelector(selectors);
            const root = await container.$(query);

            searchLog.push(createGetElementByQuerySelectorLogMessage({container, query, result: root}));

            return root;
        }

        return null;
    };

    const root = await _searchRootElement(locators.flat(), context.browser);

    if (root) {
        return root;
    }

    throw new RootElementNotFoundError(searchLog);
};
