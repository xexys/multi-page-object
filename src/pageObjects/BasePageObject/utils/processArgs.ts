import {TestContext} from '../../../contexts';
import type {Locator} from '../../../locators';
import type {PageObjectCtorLocatorArg, PageObjectCtorArgs} from '../types';
import {BasePageObject} from '../index';

type Result = {
    context?: TestContext;
    parent?: BasePageObject;
    locator: Locator;
};

export const processArgs = (po: BasePageObject, args: PageObjectCtorArgs): Result => {
    let context: TestContext | undefined;
    let parent: BasePageObject | undefined;
    let locatorArg: PageObjectCtorLocatorArg | undefined;

    if (args.length === 1) {
        if (args[0] instanceof BasePageObject) {
            parent = args[0];
        } else if (args[0] instanceof TestContext) {
            context = args[0];
        } else {
            locatorArg = args[0];
        }
    } else if (args.length === 2) {
        if (args[0] instanceof BasePageObject) {
            [parent, locatorArg] = args;
        } else {
            [context, locatorArg] = args;
        }
    }

    return {
        context,
        parent,
        locator: processLocatorArg(po, locatorArg),
    };
};

const processLocatorArg = (po: BasePageObject, arg: PageObjectCtorLocatorArg | undefined): Locator => {
    const {rootSelector} = po.getCtor();

    // Поиск по rootSelector
    if (typeof arg === 'undefined') {
        return rootSelector;
    }

    // Поиск по selector
    if (typeof arg === 'string') {
        return arg;
    }

    // Поиск по rootSelector + index
    if (typeof arg === 'number') {
        return [rootSelector, arg];
    }

    // Поиск по rootSelector + mapper
    if (typeof arg === 'function') {
        return [rootSelector, arg];
    }

    // Поиск по selector + index
    // Поиск по selector + mapper
    return arg;
};
