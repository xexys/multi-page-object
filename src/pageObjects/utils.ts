import {BasePageObject} from './BasePageObject';
import type {PageObjectCtor} from './BasePageObject';

export const isPageObjectCtor = (value: unknown): value is PageObjectCtor => {
    /**
     * @see https://stackoverflow.com/questions/14486110/how-to-check-if-a-javascript-class-inherits-another-without-creating-an-obj
     */
    return typeof value === 'function' && (value === BasePageObject || value.prototype instanceof BasePageObject);
};
