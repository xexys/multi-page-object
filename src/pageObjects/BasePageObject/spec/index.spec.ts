import fs from 'fs';
import {AssertionError} from 'assert';

import {promiseFilter} from 'shared/utils/toolkit';

import {setGlobalTestContext, UnitTestContext} from '../../../contexts';
import {getByIndex} from '../../../locators';
import type {PageElement, JSDomPageElement} from '../../../pageElements';
import {RootElementMustMatchSpecificSelectorError, RootElementNotFoundError} from '../../../errors';
import {isPageObjectRootElement} from '../utils';
import type {PageObjectCtor, PageObjectCtorLocatorArg} from '../index';
import {BasePageObject} from '../index';

setGlobalTestContext(new UnitTestContext());

/**
 * Добавим метод chain в базовый класс исключительно в целях unit-тестирования.
 *
 * Не делайте так в коде реальных тестов!
 * Используйте цепочки геттеров, как это описано в документации к модулю MultiPageObject.
 */
abstract class UnitTestPageObject<T extends Element = HTMLElement> extends BasePageObject<JSDomPageElement<T>> {
    public getRootElementValue(): Promise<T> {
        return this.rootElement.then(root => root.value);
    }

    public chain<I extends BasePageObject>(Ctor: PageObjectCtor<I>, locator?: PageObjectCtorLocatorArg): I {
        return new Ctor(this, locator);
    }
}

class Test extends UnitTestPageObject {
    static rootSelector = '*';
}

class Foo extends UnitTestPageObject {
    static rootSelector = '.foo';

    static rootCustomSelector = `${this.rootSelector}.custom`;
}

class Bar extends UnitTestPageObject {
    static rootSelector = '.bar';
}

class Baz extends UnitTestPageObject {
    static rootSelector = '.baz';
}

const innerHTML = fs.readFileSync(`${__dirname}/index.spec.html`, {encoding: 'utf-8'});

const refreshDom = () => {
    document.documentElement.innerHTML = innerHTML;
};

const runTest = async (po: UnitTestPageObject, expected: string) => {
    const root = await po.getRootElementValue();

    expect(root.textContent).toBe(expected);

    // Обновляем DOM, чтобы проверить, что цепочка вызовов не сломалась.
    refreshDom();

    expect(await po.getRootElementValue()).not.toBe(root);
    expect((await po.getRootElementValue()).textContent).toBe(expected);
};

describe('BasePageObject', () => {
    beforeAll(() => {
        refreshDom();
    });

    describe('Создал PageObject', () => {
        it('через селектор для root элемента', async () => {
            const selector = `#create-po-by-selector ${Foo.rootSelector}`;
            const foo = new Foo(selector);

            await runTest(foo, 'foo');
        });

        it('через родительский PO и индекс', async () => {
            const test = new Test('#create-po-by-parent-and-index');
            const foo = new Foo(test, 1);

            await runTest(foo, 'foo-2');
        });

        it('через родительский PO и маппер', async () => {
            const test = new Test('#create-po-by-parent-and-mapper');
            const foo = new Foo(test, async (context, elems) => {
                const result = await promiseFilter(async elem => {
                    const text = await elem.getTextContent();

                    return text?.match('foo');
                }, elems);

                return result[1];
            });

            await runTest(foo, 'foo-2');
        });

        it('через родительский PO, селектор и индекс', async () => {
            const test = new Test('#create-by-parent-and-selector-and-index');
            const foo = new Foo(test, ['div', 3]);

            await runTest(foo, 'foo-2');
        });

        it('через родительский PO, селектор и маппер', async () => {
            const test = new Test('#create-by-parent-and-selector-and-mapper');
            const foo = new Foo(test, [
                'div',
                async (context, elems) => {
                    const result = await promiseFilter(elem => isPageObjectRootElement(elem, Foo), elems);

                    return result[1];
                },
            ]);

            await runTest(foo, 'foo-2');
        });
    });

    describe('Создал вложенный PageObject', () => {
        it('через его конструктор', async () => {
            const foo = new Test('#get-po-by-ctor').chain(Foo);

            await runTest(foo, 'foo');
        });

        it('через его конструктор и индекс', async () => {
            const foo = new Test('#get-po-by-ctor-and-index').chain(Foo, 1);

            await runTest(foo, 'foo-2');
        });

        it('через его конструктор и селектор', async () => {
            const foo = new Test('#get-po-by-ctor-and-selector').chain(Foo, Foo.rootCustomSelector);

            await runTest(foo, 'foo-2');
        });

        it('через его конструктор и ручное объединение селекторов', async () => {
            const selector = `${Foo.rootSelector}:nth-child(2) ${Bar.rootSelector}:nth-child(1) ${Baz.rootSelector}:nth-child(3)`;
            const baz = new Test('#get-po-from-nested-structure').chain(Baz, selector);

            await runTest(baz, 'baz-2-1-3');
        });

        it('через его конструктор и маппер', async () => {
            const foo = new Test('#get-po-by-ctor-and-mapper').chain(Foo, getByIndex(1));

            await runTest(foo, 'foo-2');
        });

        it('через его конструктор, селектор и индекс', async () => {
            const foo = new Test('#get-po-by-ctor-and-selector-and-index').chain(Foo, ['div', 3]);

            await runTest(foo, 'foo-2');
        });

        it('через его конструктор, селектор и маппер', async () => {
            // Находит первый foo элемент с текстом красного цвета.
            const mapper = async (context: UnitTestContext, elems: PageElement[]) => {
                for (const elem of elems) {
                    const root = await isPageObjectRootElement(elem, Foo);

                    if (root && root.value.style.color === 'red') {
                        return root;
                    }
                }
                return undefined;
            };

            const foo = new Test('#get-po-by-ctor-and-selector-and-mapper').chain(Foo, ['div', mapper]);

            await runTest(foo, 'foo-2');
        });

        it('через цепочку вызова', async () => {
            const baz = new Test('#get-po-from-nested-structure').chain(Foo).chain(Bar).chain(Baz);

            await runTest(baz, 'baz-1-1-1');
        });

        it('через цепочку вызова с использованием селекторов, индексов и мапперов', async () => {
            const baz = new Test('#get-po-from-nested-structure')
                .chain(Foo, [Foo.rootSelector, 1])
                .chain(Bar, 0)
                .chain(Baz, ['div', getByIndex(2)]);

            await runTest(baz, 'baz-2-1-3');
        });

        it('через цепочку вызова и ручное объединение селекторов с применением индексов и мапперов', async () => {
            const baz = new Test('#get-po-from-nested-structure')
                .chain(Bar, [`${Foo.rootSelector} ${Bar.rootSelector}`, getByIndex(2)])
                .chain(Baz, ['.baz', 2]);

            await runTest(baz, 'baz-2-1-3');
        });
    });

    describe('Выбросил исключение если,', () => {
        it('при определении класса не задано статическое поле rootSelector', async () => {
            class PO extends BasePageObject {}

            await expect(() => new PO()).toThrowError(AssertionError);
            await expect(() => new PO()).toThrowErrorMatchingSnapshot();
        });

        it('при определении класса статическое поле rootSelector задано как falsy значение', async () => {
            class PO extends BasePageObject {
                static rootSelector = '';
            }

            await expect(() => new PO()).toThrowError(AssertionError);
            await expect(() => new PO()).toThrowErrorMatchingSnapshot();
        });

        it('при определении класса статическое поле rootSelector задано, но не является строкой', async () => {
            class PO extends BasePageObject {
                static rootSelector = {} as string;
            }

            await expect(() => new PO()).toThrowError(AssertionError);
            await expect(() => new PO()).toThrowErrorMatchingSnapshot();
        });

        it('root элемент не найден, при обращении к полю root', async () => {
            const foo = new Test('#get-po-from-nested-structure')
                .chain(Foo, [Foo.rootSelector, 1])
                .chain(Bar, 0)
                .chain(Baz, getByIndex(10));

            await expect(foo.rootElement).rejects.toThrowError(RootElementNotFoundError);
            await expect(foo.rootElement).rejects.toThrowErrorMatchingSnapshot();
        });

        it('root элемент найден, но он не соответствует root селектору', async () => {
            const rootSelector = `#root-element-must-match-specific-selector-error .foooooooo`;
            const foo = new Foo(rootSelector);

            expect(Boolean(document.querySelector(rootSelector))).toBe(true);

            await expect(foo.rootElement).rejects.toThrowError(RootElementMustMatchSpecificSelectorError);
            await expect(foo.rootElement).rejects.toThrowErrorMatchingSnapshot();
        });
    });
});
