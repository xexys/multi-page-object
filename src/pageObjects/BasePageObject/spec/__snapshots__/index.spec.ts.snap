// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasePageObject Выбросил исключение если, root элемент найден, но он не соответствует root селектору 1`] = `"Root element 'HTMLDivElement(<div class="foooooooo"></div>)' must match specific selector '.foo'!"`;

exports[`BasePageObject Выбросил исключение если, root элемент не найден, при обращении к полю root 1`] = `
"Root element not found!

    Search steps:
        - get elements by querySelectorAll (container: [object JSDomBrowser], query: '#get-po-from-nested-structure .foo', resultElementsCount: 2)
        - get element by index (index: 1, resultElement: HTMLDivElement(<div class="foo"></div>))
        - get elements by querySelectorAll (container: HTMLDivElement(<div class="foo"></div>), query: '.bar', resultElementsCount: 2)
        - get element by index (index: 0, resultElement: HTMLDivElement(<div class="bar"></div>))
        - get elements by querySelectorAll (container: HTMLDivElement(<div class="bar"></div>), query: '.baz', resultElementsCount: 3)
        - get element by mapper (mapper: 'getByIndexMapper', resultElement: undefined, debugParams: {"index":10})"
`;

exports[`BasePageObject Выбросил исключение если, при определении класса не задано статическое поле rootSelector 1`] = `"Value 'undefined' of 'PO.rootSelector' must not be empty!"`;

exports[`BasePageObject Выбросил исключение если, при определении класса статическое поле rootSelector задано как falsy значение 1`] = `"Value '' of 'PO.rootSelector' must not be empty!"`;

exports[`BasePageObject Выбросил исключение если, при определении класса статическое поле rootSelector задано, но не является строкой 1`] = `"Value '[object Object]' of 'PO.rootSelector' must be valid CSS selector string!"`;
